### ISSUE TITLE
how to run testcase serial, not in parallel by tox?

### ISSUE THREAD

**Issue by zhangzx7054** [tox-dev/tox#590]
I'd like to run testcases in serial, but I don't know how to configure tox's settings, and don't get any answer by google.
So come here for help.

**Comment by obestwalter** [comment]
Hello @zhangzx7054, I am sorry, but I do understand your question. The behaviour that you seem to want is the default in tox: environments are run after each other (in serial) and the commands in each environment are also executed one after the other. 

Can you please describe in more detail what you mean? Ideally you would provide a link to a project where you have that problem or post a `tox.ini` here and describe what you would expect to happen and show what actually happens.

**Comment by zhangzx7054** [comment]
@obestwalter I write some testcases based on testtools. I need these testcases to run serially, that means, testcase B begin to runafter testcase A finished. But actually, I find these testcases run in parallel after I  enter "`tox -e py27`".
Now I think the reason is testr. But I don't know how to config testr's settings. Can you tell me?
In my tox.ini, the "`commands`" line is " `bash -c "TESTS_DIR=./tests/ python setup.py testr --slowest --testr-args='{posargs}'"`"

Thank you.

**Comment by obestwalter** [comment]
O.k.: you execute exactly one tox environment that also executes exactly one command. This means there is nothing that tox could parallelize (even if it would - if you used detox). Therefore whatever causes the parallel execution of tests must be happening in `testr`. I don't know what `testr` is, but this parallel execution must be a feature of that testrunner, so I would recommend looking there.

P.S. Could you provide a link to that testr project? Would be interested to know what that is.

P.P.S. the invocation can be toxified this way:

    [testenv]
    setenv = TESTS_DIR = {toxinidir}/tests
    commands = python setup.py testr --slowest --testr-args='{posargs}'

**Comment by zhangzx7054** [comment]
@obestwalter Thank you for your answer.
The link below is a page about Testr.
[https://wiki.openstack.org/wiki/Testr](url)

**Comment by obestwalter** [comment]
From that page:

> TL;DR testr will run your tests in parallel (so they go faster) and it keeps robust logs of the results.

**Comment by obestwalter** [comment]
I think we can close this now. @zhangzx7054 - if you still have questions about this you can still post them here, even if the issue is closed.
